src/Analysis/Ast/Impl/Specializations/BuiltinsSpecializations.cs

```diff
@@ -175,5 +175,17 @@ public static IMember List(IPythonInterpreter interpreter, IPythonFunctionOverlo
 
             return o?.GetPythonType().GetMember(name) ?? def;
         }
+
+        public static IMember Enumerate(IPythonModule module, IPythonFunctionOverload overload, IArgumentSet argSet, IndexSpan indexSpan) {
+            var args = argSet.Values<IMember>();
+            if (args.Count > 0) {
+                var iterator = (args[0] as IPythonIterable)?.GetIterator();
+                var itemType = (iterator?.Next ?? module.Interpreter.UnknownType).GetPythonType();
+                var intType = module.Interpreter.GetBuiltinType(BuiltinTypeId.Int);
+                var content = new[] { intType.CreateInstance(ArgumentSet.WithoutContext), itemType.CreateInstance(ArgumentSet.WithoutContext) };
+                return PythonCollectionType.CreateTuple(module, content);
```

Comment:
This isn't correct. enumerate needs to be returning a iterator of tuples, not just a single tuple.

I.e. currently `x = enumerate(foo)` makes `x` a tuple, but it shouldn't be.

Typeshed actually types it as a class, interestingly: https://github.com/python/typeshed/blob/ba520034c7e84f27d3eeab6ac92a936470132ffb/stdlib/2and3/builtins.pyi#L1066

Reply:
I guess we can try class as well since Python returns `<enumerate object at ...>` from enumerate(). so `x = enumerate(foo)` would then make `x` a class, but since it is specialized, we can mimic `enumerate object` in the class documentation.

Reply:
Maybe, the other problem is that the python docs say that `enumerate` is a function, not a class, so I'd be worried about showing it as a class rather than a function too.

Reply:
I think we may have to create `Enumerator` class like .NET does with `IEnumerator`. Then `enumerate()` would return `Enumerator` and assignment handler can handle it depending on the LHS.

Reply:
Yes, that's typing.Iterable (maybe Sequence?), combined with having it produce `Tuple[int, _T]` like in Typeshed.

Realistically just those who have `__next__` is enough to specialize iterators?

Reply:
Implemented special `PythonEnumerator`